File: app/lib/core/services/settings_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../constants/app_config.dart';

class SettingsServiceException implements Exception {
  SettingsServiceException(this.message);

  final String message;

  @override
  String toString() => 'SettingsServiceException: $message';
}

class DocumentContent {
  const DocumentContent({
    required this.title,
    required this.content,
    this.updatedAt,
  });

  final String title;
  final String content;
  final DateTime? updatedAt;

  factory DocumentContent.fromJson(Map<String, dynamic> json) {
    return DocumentContent(
      title: (json['title'] as String?) ?? '',
      content: (json['content'] as String?) ?? '',
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'] as String)
          : null,
    );
  }
}

class MessageRuntimeSettings {
  const MessageRuntimeSettings({
    required this.serverStorageMode,
    required this.contentAuditMode,
  });

  static const defaults = MessageRuntimeSettings(
    serverStorageMode: 'persist',
    contentAuditMode: 'plaintext',
  );

  final String serverStorageMode;
  final String contentAuditMode;

  bool get isRelayOnly => serverStorageMode == 'relay_only';
  bool get isPersist => !isRelayOnly;
  bool get isE2ee => contentAuditMode == 'e2ee';
  bool get isPlaintext => !isE2ee;

  String get runtimeNoticeTitle => isE2ee ? '当前配置目标：端到端加密' : '当前配置目标：明文可审计';

  String get runtimeNoticeDescription {
    if (isRelayOnly) {
      return isE2ee
          ? '服务器仅做实时转发且不保存聊天记录，按当前配置目标不应被服务端审计。'
          : '服务器仅做实时转发且不保存聊天记录，消息内容仍可被服务端审计。';
    }

    return isE2ee ? '消息会保存在服务器，按当前配置目标不应被服务端审计。' : '消息会保存在服务器，管理员可审计消息内容。';
  }

  String get messageLocateMissNotice =>
      isRelayOnly
      ? '当前模式不保存聊天记录，只能定位本地缓存中的消息'
      : '当前未加载到该消息，可能已被删除或尚未同步';

  factory MessageRuntimeSettings.fromJson(Map<String, dynamic>? json) {
    return MessageRuntimeSettings(
      serverStorageMode:
          (json?['server_storage_mode'] as String?)?.trim().isNotEmpty == true
          ? json!['server_storage_mode'] as String
          : 'persist',
      contentAuditMode:
          (json?['content_audit_mode'] as String?)?.trim().isNotEmpty == true
          ? json!['content_audit_mode'] as String
          : 'plaintext',
    );
  }
}

class GeneralSettings {
  const GeneralSettings({required this.appName, required this.messageRuntime});

  final String appName;
  final MessageRuntimeSettings messageRuntime;

  factory GeneralSettings.fromJson(Map<String, dynamic> json) {
    return GeneralSettings(
      appName: (json['app_name'] as String?) ?? '',
      messageRuntime: MessageRuntimeSettings.fromJson(
        json['message_runtime'] as Map<String, dynamic>?,
      ),
    );
  }
}

class SettingsService {
  SettingsService({http.Client? client}) : _client = client ?? http.Client();

  final http.Client _client;

  Future<GeneralSettings> fetchGeneralSettings() async {
    final uri = Uri.parse('${AppConfig.apiBaseUrl}/settings/general');
    try {
      final response = await _client.get(uri);
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data is Map<String, dynamic>) {
          return GeneralSettings.fromJson(data);
        }
      }
    } catch (_) {
      // 静默失败，回退默认值
    }

    return const GeneralSettings(
      appName: '',
      messageRuntime: MessageRuntimeSettings.defaults,
    );
  }

  Future<DocumentContent> fetchPrivacyPolicy() async {
    final uri = Uri.parse('${AppConfig.apiBaseUrl}/settings/privacy-policy');
    final response = await _client.get(uri);

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data is Map<String, dynamic>) {
        return DocumentContent.fromJson(data);
      }
      throw SettingsServiceException('隐私政策数据格式异常');
    }

    throw SettingsServiceException(
      _extractMessage(response.body) ?? '隐私政策加载失败',
    );
  }

  /// 获取应用名称（公开 API，无需 token）
  Future<String> fetchAppName() async {
    try {
      final general = await fetchGeneralSettings();
      if (general.appName.isNotEmpty) {
        return general.appName;
      }
    } catch (_) {
      // 静默失败，返回默认值
    }

    final uri = Uri.parse('${AppConfig.apiBaseUrl}/settings/app-name');
    try {
      final response = await _client.get(uri);
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        return (data['app_name'] as String?) ?? '';
      }
    } catch (_) {}
    return '';
  }

  /// 获取用户协议（公开 API，无需 token）
  Future<DocumentContent> fetchUserAgreement() async {
    final uri = Uri.parse('${AppConfig.apiBaseUrl}/settings/user-agreement');
    final response = await _client.get(uri);

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data is Map<String, dynamic>) {
        return DocumentContent.fromJson(data);
      }
      throw SettingsServiceException('用户协议数据格式异常');
    }

    throw SettingsServiceException(
      _extractMessage(response.body) ?? '用户协议加载失败',
    );
  }

  /// 获取验证码设置（公开 API，无需 token）
  Future<bool> fetchRequireCaptchaForLogin() async {
    final uri = Uri.parse('${AppConfig.apiBaseUrl}/settings/captcha');
    try {
      final response = await _client.get(uri);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        return (data['require_captcha_for_login'] as bool?) ?? false;
      }
    } catch (_) {
      // 静默失败，返回默认值
    }
    return false;
  }

  String? _extractMessage(String raw) {
    try {
      final data = jsonDecode(raw);
      if (data is Map<String, dynamic>) {
        final message = data['message'];
        if (message is String && message.isNotEmpty) {
          return message;
        }
      }
    } catch (_) {}
    return null;
  }
}

File: h5-app/src/services/settings-service.ts
import { requestJson } from '@/api/http';
import { appVersion } from '@/config/version';
import type { DocumentContent, GeneralSettings, MessageRuntimeSettings, ReleasePlatform, VersionStatus } from '@/types/settings';

import { requireToken } from './session';

export const settingsService = {
  async fetchGeneralSettings(): Promise<GeneralSettings> {
    const response = await requestJson<Record<string, unknown>>('/settings/general');
    return {
      appName: String(response.app_name ?? ''),
      messageRuntime: mapRuntime(response.message_runtime),
    };
  },

  async fetchAppName(): Promise<string> {
    const response = await requestJson<Record<string, unknown>>('/settings/app-name');
    return String(response.app_name ?? '');
  },

  async fetchPrivacyPolicy(): Promise<DocumentContent> {
    return mapDocument(await requestJson<Record<string, unknown>>('/settings/privacy-policy'));
  },

  async fetchUserAgreement(): Promise<DocumentContent> {
    return mapDocument(await requestJson<Record<string, unknown>>('/settings/user-agreement'));
  },

  async submitFeedback(params: { content: string; contact?: string }): Promise<string> {
    const response = await requestJson<Record<string, unknown>>('/feedbacks', {
      method: 'POST',
      body: JSON.stringify({
        content: params.content.trim(),
        ...(params.contact?.trim() ? { contact: params.contact.trim() } : {}),
      }),
    }, requireToken());
    return String(response.message ?? '反馈提交成功');
  },

  async fetchVersionStatus(platform = detectReleasePlatform()): Promise<VersionStatus> {
    const response = await requestJson<Record<string, unknown>>(
      `/versions/latest?platform=${platform}&channel=stable&current_version=${encodeURIComponent(appVersion)}`,
    );
    const version = response.version && typeof response.version === 'object'
      ? response.version as Record<string, unknown>
      : null;
    return {
      currentVersion: String(response.current_version ?? appVersion),
      platform,
      hasUpdate: Boolean(response.has_update) && Boolean(version),
      latestVersion: version?.version == null ? null : String(version.version),
      releaseNotes: version?.release_notes == null ? null : String(version.release_notes),
      mandatory: Boolean(version?.mandatory),
      storeUrl: version?.app_store_url == null ? null : String(version.app_store_url),
    };
  },
};

export const detectReleasePlatform = (): ReleasePlatform => {
  const agent = navigator.userAgent.toLowerCase();
  if (/iphone|ipad|ipod/.test(agent)) return 'ios';
  if (agent.includes('android')) return 'android';
  if (agent.includes('windows')) return 'windows';
  if (agent.includes('mac')) return 'macos';
  return 'linux';
};

const mapRuntime = (value: unknown): MessageRuntimeSettings => {
  const data = typeof value === 'object' && value !== null ? (value as Record<string, unknown>) : {};
  return {
    serverStorageMode: String(data.server_storage_mode ?? 'persist'),
    contentAuditMode: String(data.content_audit_mode ?? 'plaintext'),
  };
};

const mapDocument = (data: Record<string, unknown>): DocumentContent => ({
  title: String(data.title ?? ''),
  content: String(data.content ?? ''),
  updatedAt: data.updated_at == null ? null : String(data.updated_at),
});
